File: backend/src/main/java/com/codeforge/module/submission/controller/SubmissionController.java
package com.codeforge.module.submission.controller;

import com.codeforge.common.dto.ApiResponse;
import com.codeforge.common.dto.PagedResponse;
import com.codeforge.common.ratelimit.RateLimit;
import com.codeforge.module.submission.dto.request.CreateSubmissionRequest;
import com.codeforge.module.submission.dto.response.SubmissionResponse;
import com.codeforge.module.submission.service.SubmissionService;
import com.codeforge.security.UserPrincipal;
import jakarta.validation.Valid;
import lombok.RequiredArgsConstructor;
import org.springframework.data.domain.PageRequest;
import org.springframework.data.domain.Pageable;
import org.springframework.data.domain.Sort;
import org.springframework.http.HttpStatus;
import org.springframework.http.ResponseEntity;
import org.springframework.security.core.annotation.AuthenticationPrincipal;
import org.springframework.web.bind.annotation.*;

@RestController
@RequestMapping("/api/v1/submissions")
@RequiredArgsConstructor
public class SubmissionController {

    private final SubmissionService submissionService;

    @PostMapping
    @RateLimit(limit = 10, windowSeconds = 60, keyPrefix = "create_submission")
    public ResponseEntity<ApiResponse<SubmissionResponse>> createSubmission(
            @AuthenticationPrincipal UserPrincipal currentUser,
            @Valid @RequestBody CreateSubmissionRequest request
    ) {
        SubmissionResponse response = submissionService.createSubmission(currentUser, request);
        return new ResponseEntity<>(
                ApiResponse.success(response, "Submission queued for evaluation"),
                HttpStatus.ACCEPTED
        );
    }

    @GetMapping("/{id}")
    public ResponseEntity<ApiResponse<SubmissionResponse>> getSubmissionById(@PathVariable Long id) {
        SubmissionResponse response = submissionService.getSubmissionById(id);
        return ResponseEntity.ok(ApiResponse.success(response, "Submission status fetched successfully"));
    }

    @GetMapping("/my")
    public ResponseEntity<ApiResponse<PagedResponse<SubmissionResponse>>> getMySubmissions(
            @AuthenticationPrincipal UserPrincipal currentUser,
            @RequestParam(required = false) Long problemId,
            @RequestParam(defaultValue = "0") int page,
            @RequestParam(defaultValue = "10") int size
    ) {
        Pageable pageable = PageRequest.of(page, Math.min(size, 50), Sort.by("id").descending());
        PagedResponse<SubmissionResponse> response = submissionService.getMySubmissions(currentUser, problemId, pageable);
        return ResponseEntity.ok(ApiResponse.success(response, "User submissions fetched successfully"));
    }

    @GetMapping("/problem/{slug}")
    public ResponseEntity<ApiResponse<PagedResponse<SubmissionResponse>>> getSubmissionsByProblem(
            @PathVariable String slug,
            @RequestParam(defaultValue = "0") int page,
            @RequestParam(defaultValue = "10") int size
    ) {
        Pageable pageable = PageRequest.of(page, Math.min(size, 50), Sort.by("id").descending());
        PagedResponse<SubmissionResponse> response = submissionService.getSubmissionsByProblemSlug(slug, pageable);
        return ResponseEntity.ok(ApiResponse.success(response, "Problem submissions fetched successfully"));
    }
}

File: backend/src/modules/submission/submission.controller.js
import { SubmissionService } from './submission.service.js';
import { createSubmissionSchema } from './submission.schema.js';
import { sendApiResponse } from '../../common/utils/apiHelpers.js';
import { ApiError } from '../../common/errors/apiError.js';

export class SubmissionController {
  static async createSubmission(req, res, next) {
    try {
      if (!req.user) {
        throw ApiError.unauthorized('Authentication required');
      }
      const validated = createSubmissionSchema.parse(req.body);
      const result = await SubmissionService.createSubmission(BigInt(req.user.id), validated);
      return sendApiResponse(res, result, 'Submission queued for evaluation', 202);
    } catch (err) {
      next(err);
    }
  }

  static async getSubmissionById(req, res, next) {
    try {
      const id = BigInt(req.params.id);
      const result = await SubmissionService.getSubmissionById(id);
      return sendApiResponse(res, result, 'Submission status fetched successfully');
    } catch (err) {
      next(err);
    }
  }

  static async getMySubmissions(req, res, next) {
    try {
      if (!req.user) {
        throw ApiError.unauthorized('Authentication required');
      }
      const problemId = req.query.problemId ? BigInt(req.query.problemId) : null;
      const page = parseInt(req.query.page || '0', 10);
      const size = parseInt(req.query.size || '10', 10);

      const result = await SubmissionService.getMySubmissions(
        BigInt(req.user.id),
        problemId,
        page,
        size
      );
      return sendApiResponse(res, result, 'User submissions fetched successfully');
    } catch (err) {
      next(err);
    }
  }

  static async getSubmissionsByProblem(req, res, next) {
    try {
      const { slug } = req.params;
      const page = parseInt(req.query.page || '0', 10);
      const size = parseInt(req.query.size || '10', 10);

      const result = await SubmissionService.getSubmissionsByProblemSlug(slug, page, size);
      return sendApiResponse(res, result, 'Problem submissions fetched successfully');
    } catch (err) {
      next(err);
    }
  }
}
